### ISSUE TITLE
Question about semi-supervised, multi-labeled classification

### ISSUE THREAD

**Issue by lnsmith54** [Alibaba-MIIL/ASL#66]
This is nice work by you and your team.

This is not an issue but rather a question.   I am considering combining your ASL loss function with a contrastive loss for the purpose of semi-supervised, multi-labeled classification.  Do you have any experience with this?  If so, would you share your experience?

Thanks,
Leslie

**Comment by mrT23** [comment]
Hi Leslie

We don't have direct experience with contrastive loss for multi-label classification.

The closest thing we did is using ASL along with Knowledge distillation. We have been doing that extensively, with significant improvement. At first, we tried adding asymmetry also to the KD loss, but we found that plain symmetric KD works just as well. We think that it might be related to the fact that KD gives more smooth and rich predictions, in contrast to the binary ground-truth of CE\Focal loss\ASL. With smooth KD predictions, there is less negative-positive imbalancing.

Tal

**Comment by lnsmith54** [comment]
Hi Tal,

Thank you for your prompt reply.

I assume you are using a teacher/student for knowledge distillation - are you using the same model size for both?  I recall seeing work where KD can yield some improvement over the teacher and I thought at the time that it might be due to label smoothing (or a mixup type effect), which is along the lines of what you are saying.

My interest is in the situation where there's lots of unlabeled data and we want to minimize the amount of manual labeling but maximize the final multi-label model performance.  I'm also planning on doing pseudo-labeling and self-training, which is really the same as teacher/student so your experience there helps me.  I appreciate your comment.

Best regards,
Leslie

**Comment by mrT23** [comment]
We use a teacher larger than the student.

for cases where you have lots of unlabeled data and minimal labeled data, from my experience KD can be effective if you do it via ensembles (teacher is an ensemble of 2-3 models, preferably a combination of conv-nets and transformers). Of course you can add iterations and so on.

good luck with your study, will keep an eye for it if you publish it :-)

Tal

**Comment by lnsmith54** [comment]
IMHO, there are several nice papers on SSL for single label classification that could be good starting points.  The trick is figuring out which one out of several possibilities.  One of the things I like about your papers (i.e., ASL, TResNet) is that you must have tried many possibilities but narrowed them down to the 3 -5 that matters most.

BTW, are you using the teacher ensemble approach for KD with fully labeled datasets?  It seems to me that it isn't so necessary there but I could be wrong.

Leslie

**Comment by mrT23** [comment]
I agree with your intuition about teacher ensembles not very useful for fully labeled datasets.
I believe that ensemble of KD teachers is especially useful for the case you described - we start from a small tagged dataset, and try to scale to a large untagged dataset. In that case, we hope that mistakes of the different models will cancel each other, and the ensemble will be able to generalize on the untagged images, and "invent" new and useful labels. We can also update the teachers along the way, and try to make them better and better.

I have been experimented with this quite a bit, although my definition for small dataset (5M tagged images - Open Images) and large dataset (25-100M images) might be different than yours :-)

Tal

**Comment by lnsmith54** [comment]
Good talking with you.

May all your research adventures be fascinating!

Best,
Leslie
